src/transport/tcp.rs

```diff
@@ -199,40 +203,52 @@ impl Stream for TcpJetStream {
 
     fn poll(&mut self) -> Result<Async<Option<<Self as Stream>::Item>>, <Self as Stream>::Error> {
         if let Ok(ref mut stream) = self.stream.try_lock() {
-            let mut buffer = [0u8; 65535];
-            match stream.poll_read(&mut buffer) {
-                Ok(Async::Ready(0)) => Ok(Async::Ready(None)),
-                Ok(Async::Ready(len)) => {
-                    let mut v = buffer.to_vec();
-                    v.truncate(len);
-                    self.nb_bytes_read += len as u64;
-                    debug!("{} bytes read on {}", len, stream.peer_addr().unwrap());
-
-                    if let Some(interceptor) = self.packet_interceptor.as_mut() {
-                        interceptor.on_new_packet(stream.peer_addr().ok(), &v);
+            let mut result = Vec::new();
+            while result.len() <= TCP_READ_LEN {
```

Comment:
I can imagine a few edge cases :

- What happen if we read exactly 8192 bytes ? We will read 8192 bytes and we will try to read again. The next call will probably return NotReady I guess... and we will lose those 8192 bytes.
- Is it possible to read data and get Ready(0) the call after ? I guess it could be possible if the connection is closed but we haven't read everything yet. We will read first then we will get the Ready(0) and lose data.
- If we reach the max (57343) and we send back the message but more data is still ready to be read, are we sure that we will be called to read again or we will be called only when another event will occur on the stream ?

Reply:
> If we reach the max (57343) and we send back the message but more data is still ready to be read, are we sure that we will be called to read again or we will be called only when another event will occur on the stream ?
Yes, the rule with streams is simple:
Assuming a Struct A (Stream) that poll on another struct B (Stream)
1. - if B is still ready and A return Ready -> A::poll will be called again
2. - if B is not ready and A return Ready -> A::poll will be called again and then will return not ready
3. - if B is not ready and A return not Ready -> A::poll will be called whenever B is ready
4. - if B is ready and A return not Ready -> A::poll will never be called again unless the task is woken up

You are talking about case 1.